pydantic/utils.py

```diff
@@ -114,6 +137,79 @@ def almost_equal_floats(value_1: float, value_2: float, *, delta: float = 1e-8)
     return abs(value_1 - value_2) <= delta
 
 
+def copy_code(code: CodeType, **update: Any) -> CodeType:
+    if PY38:
+        return code.replace(**update)  # type: ignore
+    return CodeType(*[update.get(arg, getattr(code, arg)) for arg in CODE_ARGS])
+
+
+def generate_typed_init(
+    current_init: FunctionType,
+    fields: Dict[str, 'ModelField'],
+    defaults: Dict[str, Any],
+    annotations: Dict[str, Any],
+    compiled: bool,
+) -> FunctionType:
+    """
+    Generate typed signature for init
+    """
+    if not compiled:
+        # can perform this in python, however in cython this will cause "unknown opcode" on call
+        orig_init = current_init
+        orig_code = orig_init.__code__
+        new_init = FunctionType(
+            orig_code, orig_init.__globals__, closure=orig_init.__closure__, name=orig_init.__name__
+        )
+    else:
+        orig_init = getattr(current_init, '__origin_init__', current_init)
```

Comment:
Does this fork not work if not compiled? Just wondering why we can't use the same logic in both cases (it's totally fine if we have to, I'm just trying to understand)

Reply:
I guess this would work, but it would also add redundant call to a stack and probably may cause confusion on step into model creation while debugging. Current behavior just will lead debugger right into original __init__ and won’t add any redundant calls, so I decided to stick with it.

Reply:
I see where you are coming from.

Possible counter-point: this also means that there could be bugs that show up in the compiled version and not the un-compiled version, or vice versa. I think it would be extremely annoying to figure out what was going wrong in a reported issue if it dependended on the compiled status. Do you still think it makes sense to keep the fork given that consideration?

@samuelcolvin I'd also be interested to get your opinion here.

Reply:
Good point. Personally I don't see a problem there, but this is surely need to be considered.

Reply:
My problem here is not just about different routes, but also about different signatures:

Running the following code:
```py
from datetime import datetime
from typing import List
from pydantic import BaseModel

class User(BaseModel):
    id: int
    name = 'John Doe'
    signup_ts: datetime = None
    friends: List[int] = []

user = User()
```

With pydantic not compiled I get:
```
Traceback (most recent call last):
  File "test.py", line 14, in <module>
    user = User()
  File "/home/samuel/code/pydantic/pydantic/main.py", line 293, in __init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for User
id
  field required (type=value_error.missing)
```

But when compiled I get:
```
Traceback (most recent call last):
  File "test.py", line 14, in <module>
    user = User()
  File "pydantic/utils.py", line 169, in pydantic.utils.generate_typed_init.__init__
    orig_init(*args, **kwargs)
  File "pydantic/main.py", line 293, in pydantic.main.BaseModel.__init__
    raise validation_error
pydantic.error_wrappers.ValidationError: 1 validation error for User
id
  field required (type=value_error.missing)
```

I think it's important these two tracebacks look the same and I think the traceback should **not** have the extra `pydantic.utils.generate_typed_init.__init__...` line.

I see three possible solutions to this:
1. Find a way to re arrange this function to avoid this extra step. maybe not possible.
2. Find another work around for this problem, e.g.:
  * just modify the docstring to detail the arguments
  * implement our own custom `signature()` method that provides a sensible signature for a model, without changing the actual model code
3. Manually customise the traceback so it's always the same, and perhaps even more simplified, e.g. just

```
Traceback (most recent call last):
  File "test.py", line 14, in <module>
    user = User()
pydantic.error_wrappers.ValidationError: 1 validation error for User
id
  field required (type=value_error.missing)
```
or something.

What do you think?

Reply:
You should look here for reasons why in cpython we cannot achieve same traceback as in python:
https://stackoverflow.com/questions/59072875/how-to-clone-function-in-cython-i-getting-systemerror-unknown-opcode/59087135#59087135

Reply:
Yes, i see that, I tried this myself I got something to work see 523038348177258bf83f8f75466bdc0a3a8c34c1 or the [init-signature-sandbox](https://github.com/samuelcolvin/pydantic/compare/init-signature-sandbox?expand=1) branch. But I think that approach has other problems.

I think the best solution is our own utility function to generate the signature for an init function. That `inspect.signature()` would be the same as now, but `pydantic.inspect.signature()` (or whatever) would give the new improved signature. Would that work for the original use case?

Reply:
I don't get how exactly `pydantic.inspect.signature()` would help us to solve copying problem.

Also, your approach would break models like this:

```py

class MyModelsBase(BaseModel):
    def __init__(self, **data):
        super().__init__(self, **data)
        print(f'Initialized {self} with {data}')

class MyModel(MyModelsBase):
    a: int

MyModel(a=a)   # no output, as on MyModel creation '__init__' was not in namespace and was replaced

Reply:
> 2\. Find another work around for this problem, e.g.:
> 
> * just modify the docstring to detail the arguments
> * implement our own custom `signature()` method that provides a sensible signature for a model, without changing the actual model code

Even if we modify docstring to detail the arguments, or implement our own argument, they still will be modified on the same `__init__` of a base model, so every declared model will rewrite it and eventually it'll end up with the docstring and signature of the last model created.

Reply:
I also thought about declaring `__annotations__`, `__kwdefaults__`, `__wrapped__` and `__signature__` as python `property` so they will give signatures of current models, but that doesn't seem possible too, because `WhateverModel.__init__` would still be bound to `BaseModel` and I don't see any way how these properties could know what model's signature to give on call